### PR TITLE
Add zstd compression/decompression support

### PR DESCRIPTION
**Pull request by grufino**
Adds zstd compression, issue #22393 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by grufino** [comment]
R: @tvalentyn, @pabloem

File: sdks/python/apache_beam/io/filesystem.py [hunk 1]
```diff
@@ -166,6 +177,9 @@ def _initialize_decompressor(self):
       self._decompressor = bz2.BZ2Decompressor()
     elif self._compression_type == CompressionTypes.DEFLATE:
       self._decompressor = zlib.decompressobj()
+    elif self._compression_type == CompressionTypes.ZSTD:
+      self._decompressor = zstandard.ZstdDecompressor(
+          max_window_size=2147483648).decompressobj()
```
Comment (grufino) [inline sdks/python/apache_beam/io/filesystem.py:182]: I know this seems arbitrary (and probably is), but without this param the decompressor fails with error: 
```
zstd.ZstdError: zstd decompress error: Frame requires too much memory for decoding
```
I thought about using `DEFAULT_READ_BUFFER_SIZE` but it seems too low for this param. I can either create a constant or a new parameter if necessary, let me know.

**Comment by grufino-blackbird** [comment]
R: @ryanthompson591 , @ahmedabu98
can you please have a look?

**Comment by grufino** [comment]
R: @ryanthompson591 , @ahmedabu98
can you please have a look?

File: sdks/python/apache_beam/io/filesystem.py [hunk 2]
```diff
@@ -96,14 +102,19 @@ def mime_type(cls, compression_type, default='application/octet-stream'):
         cls.BZIP2: 'application/x-bz2',
         cls.DEFLATE: 'application/x-deflate',
         cls.GZIP: 'application/x-gzip',
+        cls.ZSTD: 'application/x-zstd',
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/filesystem.py:105]: nit: I think it's `application/zstd`
https://datatracker.ietf.org/doc/html/rfc8478 
https://datatracker.ietf.org/doc/html/rfc6648#appendix-B

File: sdks/python/apache_beam/io/filesystem.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/io/filesystem.py:182]: We should have some explanation here in a comment.
Is there a related bug/discussion with zstd  maintainers that gives this recommendation from an authoritative source?

File: sdks/python/setup.py [hunk 3]
```diff
@@ -226,6 +226,7 @@ def get_portability_package_data():
         'pytz>=2018.3',
         'requests>=2.24.0,<3.0.0',
         'typing-extensions>=3.7.0',
+        'zstandard>=0.18.0',
```
Comment (tvalentyn) [inline sdks/python/setup.py:229]:
```suggestion
        'zstandard>=0.18.0,<1',
```

**Review by tvalentyn — CHANGES_REQUESTED** [review]
Thanks so much for adding this functionality!

File: sdks/python/apache_beam/io/filesystem.py (hunk 1 above)
Comment (grufino-blackbird) [reply sdks/python/apache_beam/io/filesystem.py:182]: thank you for the suggestion, I agree and found this: https://github.com/indygreg/python-zstandard/issues/157
apparently it is related to the compression level used. What do you think about adding this as a comment? It doesn't seem like the library intends to fix this, as the issue was closed, but I still think it's better to leave the value as it seems to work as a general recommendation (for my use case I tested in 10GB+ compressed or 100GB+ decompressed files and it is working fine)

Comment (grufino) [reply sdks/python/apache_beam/io/filesystem.py:182]: thank you for the suggestion, I agree and found this: https://github.com/indygreg/python-zstandard/issues/157
apparently it is related to the compression level used. What do you think about referencing this issue as a comment? It doesn't seem like the library intends to fix this in an automated way, as the issue was closed, but I still think it's better to leave the value as it seems to work as a general recommendation (for my use case I tested in 10GB+ compressed or 100GB+ decompressed files in several difference sizes and workers from 5GB to 64GB RAM and it is working fine)

Comment (tvalentyn) [reply sdks/python/apache_beam/io/filesystem.py:182]: sure. let's do that.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/filesystem.py:182]: can we also add a unit test exercising this compression on a small payload?

Comment (grufino) [reply sdks/python/apache_beam/io/filesystem.py:186]: @tvalentyn done, for the tests I reused the filesystem_test test cases for zst, hopefully that is enough, but let me know

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
There is an error:

```
Container does not include required Beam dependencies or has conflicting dependencies. If Beam dependencies have changed, you need to regenerate base_image_requirements.txt files. See: https://s.apache.org/beam-python-requirements-generate
```

I'll do that and push a commit here.

For future changes: please don't squash reviewed and unreviewed commits via force-push, until review has finalized.

**Comment by grufino** [comment]
thank you @tvalentyn, I wasn't sure the root cause of the error, lesson learned for next contributions

**Pull request merged**